src/sql/workbench/parts/dataExplorer/common/nodeContextUtils.ts

```diff
@@ -0,0 +1,125 @@
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as os from 'os';
+import { INodeContextValue } from 'sql/workbench/parts/dataExplorer/common/nodeContext';
+import { RawContextKey, IContextKeyService, IContextKey } from 'vs/platform/contextkey/common/contextkey';
+import { Disposable } from 'vs/base/common/lifecycle';
+import { NodeType } from 'sql/workbench/parts/objectExplorer/common/nodeType';
+import { IConnectionManagementService } from 'sql/platform/connection/common/connectionManagement';
+import { ConnectionProfile } from 'sql/platform/connection/common/connectionProfile';
+import { ICapabilitiesService } from 'sql/platform/capabilities/common/capabilitiesService';
+
+export class NodeContextUtils extends Disposable {
+
+	static IsDatabaseOrServer = new RawContextKey<boolean>('isDatabaseOrServer', false);
+	static IsMssqlProvided = new RawContextKey<boolean>('isMssqlProvided', false);
+	static isDatabasesFolder = new RawContextKey<boolean>('isDatabasesFolder', false);
+	static IsServer = new RawContextKey<boolean>('isServer', false);
+	static IsWindows = new RawContextKey<boolean>('isWindows', os.platform() === 'win32');
+	static IsCloud = new RawContextKey<boolean>('isCloud', false);
+
+	private isDatabaseOrServerKey: IContextKey<boolean>;
+	private isMssqlProvidedKey: IContextKey<boolean>;
+	private isDatabaseFolderKey: IContextKey<boolean>;
+	private isServerKey: IContextKey<boolean>;
+	private isCloudKey: IContextKey<boolean>;
+
+	constructor(
+		private nodeContextValue: INodeContextValue,
+		@IContextKeyService private contextKeyService: IContextKeyService,
+		@IConnectionManagementService private connectionManagementService: IConnectionManagementService,
+		@ICapabilitiesService private capabilitiesService: ICapabilitiesService
+	) {
+		super();
+		this.bindContextKeys();
+
+		// Set additional node context keys
+		if (this.nodeContextValue.node) {
+			const node = this.nodeContextValue.node;
+			if (node.payload) {
+				this.isDatabasesFolder();
+				this.isProvidedByMssql();
+				this.isCloud();
+			}
+			if (node.type) {
+				this.isServer();
+			}
+			if (node.contextValue) {
+				this.isDatabaseOrServer();
+			}
+		}
+	}
+
+	private bindContextKeys(): void {
+		this.isDatabaseOrServerKey = NodeContextUtils.IsDatabaseOrServer.bindTo(this.contextKeyService);
+		this.isMssqlProvidedKey = NodeContextUtils.IsMssqlProvided.bindTo(this.contextKeyService);
+		this.isDatabaseFolderKey = NodeContextUtils.isDatabasesFolder.bindTo(this.contextKeyService);
+		this.isServerKey = NodeContextUtils.IsServer.bindTo(this.contextKeyService);
+		this.isCloudKey = NodeContextUtils.IsCloud.bindTo(this.contextKeyService);
+	}
+
+	/**
+	 * Helper function to tell whether node is a database or server or not
+	 */
+	private isDatabaseOrServer(): void {
```

Comment:
Why do you have to check both node.type and node.contextValue? Why wouldn't we just have a single one that we use for this and expect people to fill in?

Reply:
Because one could be an ObjectExplorerNode and one could be a node coming from the extension

Reply:
Why not just update OE to also fill in the type so you don't have to do both these checks and instead can just have type be a key on the node?

Reply:
The OE doesn't know about the node yet. It gets the node context from here and then goes to the shim.